052-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,167 @@
+
+In Kafka, metric reporters are configured per Kafka client or broker. This makes it hard to use reporters in client-side applications that use multiple clients such as Connect, MirrorMaker, Streams but also Strimzi components like the HTTP bridge. At the moment, for these components it's preferable to keep the current metrics pipeline with JMX and jmx_exporter. For these two reasons, I propose only using reporters for brokers/controllers and only using them in KRaft mode.
+
+## Affected/not affected projects
+
+This affects `strimzi-kafka-operator `. Reporters are not usable with ZooKeeper and currently don't work well with client side application like Kafka Connect, MirrorMaker and the HTTP bridge. So Strimzi should keep support for jmx_exporter. We may reconsider dropping support once ZooKeeper is not supported anymore and if improvements in Apache Kafka make reporters usable with client side applications.
+
+The reporter will be a new project and should be usable without Strimzi. As it's new code, it should be first release as a preview feature (disabled by default). Once KRaft becomes the preferred deployment option, we should make the reporter the recommended option to expose metrics for brokers and controllers.
+
+In `strimzi-kafka-operator`, we will need new examples demonstrating how to use the reporter, new grafana dashboards with metric names from the reporter.
+
+Due to the way the OAuth plugin exports its metrics, the Prometheus reporter won't be usable with the plugin. It will require [KIP-877](https://cwiki.apache.org/confluence/display/KAFKA/KIP-877%3A+Mechanism+for+plugins+and+connectors+to+register+metrics) to work.
+
+## Compatibility
+
+Differences with jmx_exporter metrics:
+
+- The reporter does not compute 1/5/15 minute rate, mean, max, min, stddev metrics. It's preferable to compute them in Prometheus instead of on the client side.
+- The reporter is missing the `kafka_server_app_info_starttimems` metric with the client/broker id label. (Due to [KAFKA-15186](https://issues.apache.org/jira/browse/KAFKA-15186))
+
+Assuming jmx_exporter does not have any rules, this is the other main metric change:
+
+- With the reporter, the `name` field is put directly in the metric name. For example this MBean, `kafka.server:type=ZooKeeperClientMetrics,name=ZooKeeperRequestLatencyMs` with the `Count` attribute is converted to `kafka_server_zookeeperclientmetrics_zookeeperrequestlatencyms_count`. By default jmx_exporter keeps the name as a label, `kafka_server_zookeeperclientmetrics_count{name="ZooKeeperRequestLatencyMs",}`.
+
+### Comparison
+
+Actually with the example rules from [kafka-metrics.yaml](https://github.com/strimzi/strimzi-kafka-operator/blob/main/examples/metrics/kafka-metrics.yaml), jmx_exporter generates metric names very similar to the reporter.
```

Comment:
What does very similar actually mean though? I'm specifically interested in whether the example dashboards will get broken by any changes.

Reply:
Yes some dashboards will be broken. The section just below highlights the differences between the 2 outputs.